Bricks/template/typelist.h

```diff
@@ -0,0 +1,254 @@
+static_assert(std::is_same<TypeListImpl<int>, Deduplicate<TypeListImpl<int, int>>>::value, "");
+static_assert(std::is_same<TypeListImpl<int>, Deduplicate<TypeListImpl<int, int, int>>>::value, "");
+static_assert(std::is_same<TypeListImpl<int, double>, Deduplicate<TypeListImpl<int, double>>>::value, "");
+static_assert(std::is_same<TypeListImpl<int, double>,
+                           Deduplicate<TypeListImpl<int, int, int, double, double, double>>>::value,
+              "");
+static_assert(std::is_same<TypeListImpl<int, double>, Deduplicate<TypeListImpl<int, double, int>>>::value, "");
+static_assert(std::is_same<TypeListImpl<double, int>, Deduplicate<TypeListImpl<double, int, double>>>::value,
+              "");
+
+// User-facing `TypeList<TS...>` creates a type list that contains the union of all passed in types,
+// performing flattening and deduplication of all input types.
+template <typename... TS>
+using TypeList = Deduplicate<Flatten<TypeListImpl<TS...>>>;
+
+static_assert(std::is_same<TypeListImpl<>, TypeList<>>::value, "");
+static_assert(std::is_same<TypeListImpl<int>, TypeList<int>>::value, "");
+static_assert(std::is_same<TypeListImpl<int>, TypeList<TypeList<TypeList<int>>>>::value, "");
+static_assert(std::is_same<TypeListImpl<int>, TypeList<int, int>>::value, "");
+static_assert(std::is_same<TypeListImpl<int>, TypeList<int, TypeList<int>>>::value, "");
+static_assert(std::is_same<TypeListImpl<int>, TypeList<int, TypeList<TypeList<int>>>>::value, "");
+static_assert(std::is_same<TypeListImpl<int, double>, TypeList<int, TypeList<TypeList<int>, double>>>::value,
+              "");
+static_assert(std::is_same<TypeListImpl<int, double>, TypeList<int, int, double, double>>::value, "");
+static_assert(std::is_same<TypeListImpl<int, double>, TypeList<int, double, int>>::value, "");
+static_assert(std::is_same<TypeListImpl<int, double>, TypeList<int, double, int, double>>::value, "");
+static_assert(std::is_same<TypeListImpl<int, double>, TypeList<TypeList<int, double, int, double>>>::value, "");
+
+static_assert(!std::is_same<TypeListImpl<int>, TypeList<double>>::value, "");
+static_assert(!std::is_same<TypeListImpl<int, double>, TypeList<double, int>>::value, "");
+
+template <typename... TS>
+struct IsTypeList {
+  enum { value = false };
+};
+
+template <typename... TS>
+struct IsTypeList<TypeListImpl<TS...>> {
+  enum { value = true };
+};
+
+static_assert(IsTypeList<TypeListImpl<>>::value, "");
+static_assert(IsTypeList<TypeListImpl<int>>::value, "");
+static_assert(IsTypeList<TypeListImpl<int, double>>::value, "");
+static_assert(!IsTypeList<int>::value, "");
+static_assert(!IsTypeList<double>::value, "");
+
+}  // namespace metaprogramming
+}  // namespace bricks
+
+// Export some symbols into global scope.
+using bricks::metaprogramming::TypeList;
+using bricks::metaprogramming::TypeListContains;
+using bricks::metaprogramming::IsTypeList;
+using bricks::metaprogramming::TypeListSize;
```

Comment:
Do we now have a policy to inject methods from `*.h`-s into global scope? Or it's a special case?